(bug) Panic on applyScheduledRolloutSteps

### Observed behaviour

Hello!

We encountered a panic due to concurrent map writes in the production environment when using flags with ScheduledRollout:

```
fatal error: concurrent map writes

.../internal/flag.(*InternalFlag).applyScheduledRolloutSteps(0xc154c31590)
.../internal/flag/internal_flag.go:195 +0x4fc
.../internal/flag.(*InternalFlag).Value(0xc154c31590, {0x1a9e1ef, 0xb}, {0x1d9b470, 0xc18779d020}, {0xc19d83f5f0?, {0x16ea860?, 0x2aa92c0?}})
```
(line numbers may differ from the current release)

The problem is reproduced on the current main branch (go 1.22):
```
=== CONT  TestInternalFlag_ApplySheduledRollout/scheduledRollout#131
    testing.go:1398: race detected during execution of test
--- FAIL: TestInternalFlag_ApplySheduledRollout/scheduledRollout#131 (0.01s)
==================
WARNING: DATA RACE
Write at 0x00c000197140 by goroutine 784:
  runtime.mapassign_faststr()
      /usr/local/go/src/runtime/map_faststr.go:203 +0x0
  github.com/thomaspoignant/go-feature-flag/internal/flag.(*InternalFlag).applyScheduledRolloutSteps()
      /home/xxx/projects/other/go-feature-flag/internal/flag/internal_flag.go:192 +0xa79
  github.com/thomaspoignant/go-feature-flag/internal/flag.(*InternalFlag).Value()
      /home/xxx/projects/other/go-feature-flag/internal/flag/internal_flag.go:65 +0xec
  github.com/thomaspoignant/go-feature-flag/internal/flag_test.TestInternalFlag_ApplySheduledRollout.func1()
      /home/xxx/projects/other/go-feature-flag/internal/flag/internal_flag_test.go:63 +0x2da
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1742 +0x44
```

### Expected Behavior

Concurrent access to a flag value should not cause panic

### Steps to reproduce

Reproduced with `-race` argument in a small number of cases (10) and without `-race` in a large number of cases (>1000)

```golang
func TestInternalFlag_ApplySheduledRollout(t *testing.T) {
	cases := 1000
	internalFlag := flag.InternalFlag{
		Variations: &map[string]*interface{}{
			"variation_A": testconvert.Interface("value_A"),
			"variation_B": testconvert.Interface("value_B"),
		},
		DefaultRule: &flag.Rule{
			VariationResult: testconvert.String("variation_A"),
		},
		Metadata: &map[string]interface{}{
			"description": "this is a flag",
			"issue-link":  "https://issue.link/GOFF-1",
		},
		Scheduled: &[]flag.ScheduledStep{
			{
				InternalFlag: flag.InternalFlag{
					DefaultRule: &flag.Rule{
						VariationResult: testconvert.String("variation_B"),
					},
				},
				Date: testconvert.Time(time.Now().Add(-2 * time.Second)),
			},
			{
				InternalFlag: flag.InternalFlag{
					Variations: &map[string]*interface{}{
						"variation_B": testconvert.Interface("value_QWERTY"),
					},
				},
				Date: testconvert.Time(time.Now().Add(-1 * time.Second)),
			},
		},
	}
	for i := 0; i < cases; i++ {
		t.Run("scheduledRollout", func(t *testing.T) {
			t.Parallel()

			wantResult := "value_QWERTY"
			wantDetails := flag.ResolutionDetails{
				Variant: "variation_B",
				Reason:  flag.ReasonStatic,
				Metadata: map[string]interface{}{
					"description": "this is a flag",
					"issue-link":  "https://issue.link/GOFF-1",
				},
			}

			got, got1 := internalFlag.Value(
				"flag",
				ffcontext.NewEvaluationContextBuilder("user-key").Build(),
				flag.Context{
					DefaultSdkValue: "value_default",
				},
			)
			assert.Equalf(t, wantResult, got, "not expected value: %s", cmp.Diff(wantResult, got))
			assert.Equalf(t, wantDetails, got1, "not expected value: %s", cmp.Diff(wantDetails, got1))
		})
	}
}
```

--- a/internal/flag/internal_flag.go
+++ b/internal/flag/internal_flag.go
@@ -1,6 +1,7 @@
 package flag
 
 import (
+	"encoding/json"
 	"fmt"
 	"github.com/thomaspoignant/go-feature-flag/ffcontext"
 	"maps"
@@ -62,39 +63,47 @@ func (f *InternalFlag) Value(
 	flagContext Context,
 ) (interface{}, ResolutionDetails) {
 	evaluationDate := DateFromContextOrDefault(evaluationCtx, time.Now())
-	f.applyScheduledRolloutSteps(evaluationDate)
+	flag, err := f.applyScheduledRolloutSteps(evaluationDate)
+
+	if err != nil {
+		return flagContext.DefaultSdkValue, ResolutionDetails{
+			Variant:   VariationSDKDefault,
+			Reason:    ReasonError,
+			ErrorCode: ErrorCodeGeneral,
+		}
+	}
 
 	if flagContext.EvaluationContextEnrichment != nil {
 		maps.Copy(evaluationCtx.GetCustom(), flagContext.EvaluationContextEnrichment)
 	}
 
-	if f.IsDisable() || f.isExperimentationOver(evaluationDate) {
+	if flag.IsDisable() || flag.isExperimentationOver(evaluationDate) {
 		return flagContext.DefaultSdkValue, ResolutionDetails{
 			Variant:   VariationSDKDefault,
 			Reason:    ReasonDisabled,
-			Cacheable: f.isCacheable(),
-			Metadata:  f.GetMetadata(),
+			Cacheable: flag.isCacheable(),
+			Metadata:  flag.GetMetadata(),
 		}
 	}
 
-	variationSelection, err := f.selectVariation(flagName, evaluationCtx)
+	variationSelection, err := flag.selectVariation(flagName, evaluationCtx)
 	if err != nil {
 		return flagContext.DefaultSdkValue,
 			ResolutionDetails{
 				Variant:   VariationSDKDefault,
 				Reason:    ReasonError,
 				ErrorCode: ErrorFlagConfiguration,
-				Metadata:  f.GetMetadata(),
+				Metadata:  flag.GetMetadata(),
 			}
 	}
 
-	return f.GetVariationValue(variationSelection.name), ResolutionDetails{
+	return flag.GetVariationValue(variationSelection.name), ResolutionDetails{
 		Variant:   variationSelection.name,
 		Reason:    variationSelection.reason,
 		RuleIndex: variationSelection.ruleIndex,
 		RuleName:  variationSelection.ruleName,
 		Cacheable: variationSelection.cacheable,
-		Metadata:  f.GetMetadata(),
+		Metadata:  flag.GetMetadata(),
 	}
 }
 
@@ -170,47 +179,62 @@ func (f *InternalFlag) selectVariation(flagName string, ctx ffcontext.Context) (
 // nolint: gocognit
 // applyScheduledRolloutSteps is checking if the flag has a scheduled rollout configured.
 // If yes we merge the changes to the current flag.
-func (f *InternalFlag) applyScheduledRolloutSteps(evaluationDate time.Time) {
-	if f.Scheduled != nil {
-		for _, steps := range *f.Scheduled {
-			if steps.Date != nil && (steps.Date.Before(evaluationDate) || steps.Date.Equal(evaluationDate)) {
-				f.Rules = MergeSetOfRules(f.GetRules(), steps.GetRules())
-				if steps.Disable != nil {
-					f.Disable = steps.Disable
-				}
+func (f *InternalFlag) applyScheduledRolloutSteps(evaluationDate time.Time) (*InternalFlag, error) {
+	if f.Scheduled == nil {
+		return f, nil
+	}
 
-				if steps.TrackEvents != nil {
-					f.TrackEvents = steps.TrackEvents
-				}
+	// We are doing a deep copy the flag to avoid modifying the original flag.
+	// The deep copy is done to fix this issue https://github.com/thomaspoignant/go-feature-flag/issues/2256
+	data, err := json.Marshal(f)
+	if err != nil {
+		return &InternalFlag{}, err
+	}
+	var flagCopy *InternalFlag
+	if err := json.Unmarshal(data, &flagCopy); err != nil {
+		return &InternalFlag{}, err
+	}
 
-				if steps.DefaultRule != nil {
-					f.DefaultRule.MergeRules(*steps.DefaultRule)
-				}
+	// We apply the scheduled rollout
+	for _, steps := range *f.Scheduled {
+		if steps.Date != nil && (steps.Date.Before(evaluationDate) || steps.Date.Equal(evaluationDate)) {
+			flagCopy.Rules = MergeSetOfRules(f.GetRules(), steps.GetRules())
+			if steps.Disable != nil {
+				flagCopy.Disable = steps.Disable
+			}
 
-				if steps.Variations != nil {
-					for key, value := range steps.GetVariations() {
-						f.GetVariations()[key] = value
-					}
-				}
+			if steps.TrackEvents != nil {
+				flagCopy.TrackEvents = steps.TrackEvents
+			}
 
-				if steps.Version != nil {
-					f.Version = steps.Version
-				}
+			if steps.DefaultRule != nil {
+				flagCopy.DefaultRule.MergeRules(*steps.DefaultRule)
+			}
 
-				if steps.Experimentation != nil {
-					if f.Experimentation == nil {
-						f.Experimentation = &ExperimentationRollout{}
-					}
-					if steps.Experimentation.Start != nil {
-						f.Experimentation.End = steps.Experimentation.End
-					}
-					if steps.Experimentation.End != nil {
-						f.Experimentation.End = steps.Experimentation.End
-					}
+			if steps.Variations != nil {
+				for key, value := range steps.GetVariations() {
+					flagCopy.GetVariations()[key] = value
+				}
+			}
+
+			if steps.Version != nil {
+				flagCopy.Version = steps.Version
+			}
+
+			if steps.Experimentation != nil {
+				if flagCopy.Experimentation == nil {
+					flagCopy.Experimentation = &ExperimentationRollout{}
+				}
+				if steps.Experimentation.Start != nil {
+					flagCopy.Experimentation.End = steps.Experimentation.End
+				}
+				if steps.Experimentation.End != nil {
+					flagCopy.Experimentation.End = steps.Experimentation.End
 				}
 			}
 		}
 	}
+	return flagCopy, nil
 }
 
 // isExperimentationOver checks if we are in an experimentation or not

--- a/internal/flag/internal_flag_test.go
+++ b/internal/flag/internal_flag_test.go
@@ -2201,3 +2201,62 @@ func TestInternalFlag_IsValid(t *testing.T) {
 		})
 	}
 }
+func TestInternalFlag_ApplySheduledRollout(t *testing.T) {
+	cases := 1000
+	internalFlag := flag.InternalFlag{
+		Variations: &map[string]*interface{}{
+			"variation_A": testconvert.Interface("value_A"),
+			"variation_B": testconvert.Interface("value_B"),
+		},
+		DefaultRule: &flag.Rule{
+			VariationResult: testconvert.String("variation_A"),
+		},
+		Metadata: &map[string]interface{}{
+			"description": "this is a flag",
+			"issue-link":  "https://issue.link/GOFF-1",
+		},
+		Scheduled: &[]flag.ScheduledStep{
+			{
+				InternalFlag: flag.InternalFlag{
+					DefaultRule: &flag.Rule{
+						VariationResult: testconvert.String("variation_B"),
+					},
+				},
+				Date: testconvert.Time(time.Now().Add(-2 * time.Second)),
+			},
+			{
+				InternalFlag: flag.InternalFlag{
+					Variations: &map[string]*interface{}{
+						"variation_B": testconvert.Interface("value_QWERTY"),
+					},
+				},
+				Date: testconvert.Time(time.Now().Add(-1 * time.Second)),
+			},
+		},
+	}
+	for i := 0; i < cases; i++ {
+		t.Run("scheduledRollout", func(t *testing.T) {
+			t.Parallel()
+
+			wantResult := "value_QWERTY"
+			wantDetails := flag.ResolutionDetails{
+				Variant: "variation_B",
+				Reason:  flag.ReasonStatic,
+				Metadata: map[string]interface{}{
+					"description": "this is a flag",
+					"issue-link":  "https://issue.link/GOFF-1",
+				},
+			}
+
+			got, got1 := internalFlag.Value(
+				"flag",
+				ffcontext.NewEvaluationContextBuilder("user-key").Build(),
+				flag.Context{
+					DefaultSdkValue: "value_default",
+				},
+			)
+			assert.Equalf(t, wantResult, got, "not expected value: %s", cmp.Diff(wantResult, got))
+			assert.Equalf(t, wantDetails, got1, "not expected value: %s", cmp.Diff(wantDetails, got1))
+		})
+	}
+}

--- a/internal/flag/internal_flag_test.go
+++ b/internal/flag/internal_flag_test.go
@@ -1663,6 +1663,43 @@ func TestInternalFlag_Value(t *testing.T) {
 				Cacheable: true,
 			},
 		},
+		{
+			name: "Should return sdk default value when we have an error in the deep copy",
+			flag: flag.InternalFlag{
+				Experimentation: &flag.ExperimentationRollout{
+					Start: testconvert.Time(time.Now().Add(-15 * time.Second)),
+					End:   testconvert.Time(time.Now().Add(-5 * time.Second)),
+				},
+				Metadata: &map[string]interface{}{
+					"description": make(chan int),
+					"issue-link":  "https://issue.link/GOFF-1",
+				},
+				Scheduled: &[]flag.ScheduledStep{
+					{
+						Date: testconvert.Time(time.Now().Add(-10 * time.Second)),
+						InternalFlag: flag.InternalFlag{
+							Experimentation: &flag.ExperimentationRollout{
+								Start: testconvert.Time(time.Now().Add(-5 * time.Second)),
+								End:   testconvert.Time(time.Now().Add(5 * time.Second)),
+							},
+						},
+					},
+				},
+			},
+			args: args{
+				flagName: "my-flag",
+				user:     ffcontext.NewEvaluationContext("user-key"),
+				flagContext: flag.Context{
+					DefaultSdkValue: "default-sdk",
+				},
+			},
+			want: "default-sdk",
+			want1: flag.ResolutionDetails{
+				Variant:   "SdkDefault",
+				Reason:    flag.ReasonError,
+				ErrorCode: flag.ErrorCodeGeneral,
+			},
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

